### ISSUE TITLE
Non-destructive (reeditable filters)

### ISSUE THREAD

**Issue by LinuxBeaver** [GreycLab/gmic#372]
I am wondering if GMIC has plans to implement non-destructive editing where the original image can be restored and filters can be re-edited. 

If GMIC saved an copy of a layer before edits and reloaded it on command, non-destructive editing would be possible as long as the GMIC window is kept upon. The filter would continuously reapply to the original back up layer. Other open source raster editors like Krita and Pixelitor do this under the names (filter layers and smart objects) It would definitely take the possibility of the software to the next level if this were possible. What are the GMIC teams thoughts? I think it would be possible if we had a smart/active filter check box.

**Comment by dtschump** [comment]
If such a feature is implemented, it must be done directly by the host application, not by G'MIC itself.
Technically speaking, It could be possible to associate a G'MIC command to a "filter layer" and let the user change it if needed. But this is clearly something related to the interface of the host software, not to the processing engine strictly speaking.

**Comment by 0xC0000054** [comment]
> If such a feature is implemented, it must be done directly by the host application, not by G'MIC itself.

I implemented non-destructive editing support for the 8bf plugin when running under Photoshop, the plugin can can be used as a [Smart Filter](https://helpx.adobe.com/photoshop/using/applying-smart-filters.html) in Photoshop CS3 (version 10.0) and later.
This was part of some work improving compatibility between the G'MIC 8bf plugin and Photoshop's Actions/Scripting system. 
The required Photoshop APIs were added in version 5.0, so other 3rd-party hosts could use them to implement a similar feature but I am not aware of any that do so.

The G'MIC-Qt APIs that allow a host to get and set the filter parameters were added in version 2.9.8.
